autoPyTorch/api/base_task.py

```diff
@@ -1,139 +1,740 @@
-        phase after finding a best hyperparameter configuration after the hpo
-        phase.
+        """
+        Refit all models found with fit to new data.
+
+        Necessary when using cross-validation. During training, autoPyTorch
+        fits each model k times on the dataset, but does not keep any trained
+        model and can therefore not be used to predict for new data points.
+        This methods fits all models found during a call to fit on the data
+        given. This method may also be used together with holdout to avoid
+        only using 66% of the training data to fit the final model.
         Args:
             dataset: (Dataset)
                 The argument that will provide the dataset splits. It can either
                 be a dictionary with the splits, or the dataset object which can
                 generate the splits based on different restrictions.
-            model_config: (Configuration)
-                The configuration of the model.
+            budget_config: (Optional[Dict[str, Union[int, str]]])
+                can contain keys from 'budget_type' and the budget
+                specified using 'epochs' or 'runtime'.
         Returns:
-            Value of the evaluation metric calculated on the test set.
+            self
         """
-        # TODO the model configuration is for the pipeline
-        # instead of it being given from the hpo algorithm
-        # it takes it from us ?
 
-        dataset_properties = dataset.get_dataset_properties()
+        self._logger = self._get_logger(dataset.dataset_name)
+
+        dataset_requirements = get_dataset_requirements(
+            info=self._get_required_dataset_properties(dataset))
+        dataset_properties = dataset.get_dataset_properties(dataset_requirements)
+
+        X: Dict[str, Any] = dict({'dataset_properties': dataset_properties,
+                                  'backend': self._backend,
+                                  })
+        X.update({**self.default_pipeline_options, **budget_config})
+        if self.models_ is None or len(self.models_) == 0 or self.ensemble_ is None:
+            self._load_models(dataset.resampling_strategy)
+
+        # Refit is not applicable when ensemble_size is set to zero.
+        if self.ensemble_ is None:
+            raise ValueError("Refit can only be called if 'ensemble_size != 0'")
+
+        for identifier in self.models_:
+            model = self.models_[identifier]
+            # this updates the model inplace, it can then later be used in
+            # predict method
 
-        X = {}
-        X.update(dataset_properties)
-        X.update(self.pipeline_config)
+            # try to fit the model. If it fails, shuffle the data. This
+            # could alleviate the problem in algorithms that depend on
+            # the ordering of the data.
+            fit_and_suppress_warnings(self._logger, model, X, y=None)
 
-        self.pipeline.set_hyperparameters(model_config)
```

Comment:
Could we add a new method that takes a dataset and a model config (a sample from a ConfigSpace object) and trains the pipeline? Basically what l. 129/130 does. That is a pretty useful feature if you just want to evaluate one config as a user. Maybe call that one `.fit` or `.train`